File: src/main/java/com/github/cybellereaper/wizpets/core/persistence/PetStorage.java
package com.github.cybellereaper.wizpets.core.persistence;

import com.github.cybellereaper.wizpets.api.PetRecord;
import com.github.cybellereaper.wizpets.api.StatSet;
import com.github.cybellereaper.wizpets.api.persistence.PetPersistence;
import com.google.common.collect.ImmutableList;
import io.vavr.control.Option;
import java.util.List;
import java.util.Optional;
import java.util.function.Function;
import lombok.NonNull;
import org.bukkit.NamespacedKey;
import org.bukkit.persistence.PersistentDataAdapterContext;
import org.bukkit.persistence.PersistentDataContainer;
import org.bukkit.persistence.PersistentDataHolder;
import org.bukkit.persistence.PersistentDataType;
import org.bukkit.plugin.java.JavaPlugin;
import org.jooq.lambda.Seq;

public final class PetStorage implements PetPersistence {
  private final NamespacedKey rootKey;
  private final NamespacedKey nameKey;
  private final NamespacedKey generationKey;
  private final NamespacedKey breedKey;
  private final NamespacedKey talentsKey;
  private final NamespacedKey mountKey;
  private final NamespacedKey flightKey;
  private final NamespacedKey evHealthKey;
  private final NamespacedKey evAttackKey;
  private final NamespacedKey evDefenseKey;
  private final NamespacedKey evMagicKey;
  private final NamespacedKey ivHealthKey;
  private final NamespacedKey ivAttackKey;
  private final NamespacedKey ivDefenseKey;
  private final NamespacedKey ivMagicKey;

  public PetStorage(@NonNull JavaPlugin plugin) {
    rootKey = new NamespacedKey(plugin, "pet");
    nameKey = new NamespacedKey(plugin, "name");
    generationKey = new NamespacedKey(plugin, "generation");
    breedKey = new NamespacedKey(plugin, "breed_count");
    talentsKey = new NamespacedKey(plugin, "talents");
    mountKey = new NamespacedKey(plugin, "mount_unlocked");
    flightKey = new NamespacedKey(plugin, "flight_unlocked");
    evHealthKey = new NamespacedKey(plugin, "ev_health");
    evAttackKey = new NamespacedKey(plugin, "ev_attack");
    evDefenseKey = new NamespacedKey(plugin, "ev_defense");
    evMagicKey = new NamespacedKey(plugin, "ev_magic");
    ivHealthKey = new NamespacedKey(plugin, "iv_health");
    ivAttackKey = new NamespacedKey(plugin, "iv_attack");
    ivDefenseKey = new NamespacedKey(plugin, "iv_defense");
    ivMagicKey = new NamespacedKey(plugin, "iv_magic");
  }

  @Override
  public Optional<PetRecord> load(@NonNull PersistentDataHolder holder) {
    return decodeFromParent(holder.getPersistentDataContainer());
  }

  @Override
  public void save(@NonNull PersistentDataHolder holder, @NonNull PetRecord record) {
    PersistentDataContainer parent = holder.getPersistentDataContainer();
    PersistentDataContainer container = encode(parent.getAdapterContext(), record);
    parent.set(rootKey, PersistentDataType.TAG_CONTAINER, container);
  }

  @Override
  public void clear(@NonNull PersistentDataHolder holder) {
    holder.getPersistentDataContainer().remove(rootKey);
  }

  @Override
  public boolean exists(@NonNull PersistentDataHolder holder) {
    return holder.getPersistentDataContainer().has(rootKey, PersistentDataType.TAG_CONTAINER);
  }

  @Override
  public Optional<PetRecord> compute(
      @NonNull PersistentDataHolder holder,
      @NonNull Function<Optional<PetRecord>, Optional<PetRecord>> operation) {
    PersistentDataContainer parent = holder.getPersistentDataContainer();
    Optional<PetRecord> current = decodeFromParent(parent);
    Optional<PetRecord> result = operation.apply(current);
    if (result == null) {
      throw new NullPointerException("Operation must not return null");
    }
    if (result.isPresent()) {
      save(holder, result.get());
    } else {
      clear(holder);
    }
    return result;
  }

  @Override
  public PersistentDataContainer encode(
      @NonNull PersistentDataAdapterContext context, @NonNull PetRecord record) {
    PersistentDataContainer container = context.newPersistentDataContainer();

    container.set(nameKey, PersistentDataType.STRING, record.displayName());
    container.set(generationKey, PersistentDataType.INTEGER, record.generation());
    container.set(breedKey, PersistentDataType.INTEGER, record.breedCount());
    container.set(talentsKey, PersistentDataType.STRING, String.join(";", record.talentIds()));
    container.set(mountKey, PersistentDataType.BYTE, asByte(record.mountUnlocked()));
    container.set(flightKey, PersistentDataType.BYTE, asByte(record.flightUnlocked()));

    container.set(evHealthKey, PersistentDataType.DOUBLE, record.evs().health());
    container.set(evAttackKey, PersistentDataType.DOUBLE, record.evs().attack());
    container.set(evDefenseKey, PersistentDataType.DOUBLE, record.evs().defense());
    container.set(evMagicKey, PersistentDataType.DOUBLE, record.evs().magic());

    container.set(ivHealthKey, PersistentDataType.DOUBLE, record.ivs().health());
    container.set(ivAttackKey, PersistentDataType.DOUBLE, record.ivs().attack());
    container.set(ivDefenseKey, PersistentDataType.DOUBLE, record.ivs().defense());
    container.set(ivMagicKey, PersistentDataType.DOUBLE, record.ivs().magic());

    return container;
  }

  @Override
  public Optional<PetRecord> decode(@NonNull PersistentDataContainer container) {
    return Option.of(container.get(nameKey, PersistentDataType.STRING))
        .filter(name -> !name.isBlank())
        .map(name -> buildRecord(container, name))
        .toJavaOptional();
  }

  private Optional<PetRecord> decodeFromParent(PersistentDataContainer parent) {
    return Option.of(parent.get(rootKey, PersistentDataType.TAG_CONTAINER))
        .flatMap(container -> Option.ofOptional(decode(container)))
        .toJavaOptional();
  }

  private PetRecord buildRecord(PersistentDataContainer container, String name) {
    List<String> talents =
        Option.of(container.get(talentsKey, PersistentDataType.STRING))
            .filter(raw -> !raw.isBlank())
            .map(
                raw ->
                    Seq.of(raw.split(";"))
                        .map(String::trim)
                        .filter(trimmed -> !trimmed.isEmpty())
                        .toList())
            .map(ImmutableList::copyOf)
            .getOrElse(ImmutableList::of);

    StatSet evs =
        new StatSet(
            getDouble(container, evHealthKey),
            getDouble(container, evAttackKey),
            getDouble(container, evDefenseKey),
            getDouble(container, evMagicKey));

    StatSet ivs =
        new StatSet(
            getDouble(container, ivHealthKey),
            getDouble(container, ivAttackKey),
            getDouble(container, ivDefenseKey),
            getDouble(container, ivMagicKey));

    int generation =
        Option.of(container.get(generationKey, PersistentDataType.INTEGER)).getOrElse(1);
    int breedCount = Option.of(container.get(breedKey, PersistentDataType.INTEGER)).getOrElse(0);
    boolean mountUnlocked =
        Option.of(container.get(mountKey, PersistentDataType.BYTE))
            .exists(value -> value == (byte) 1);
    boolean flightUnlocked =
        Option.of(container.get(flightKey, PersistentDataType.BYTE))
            .exists(value -> value == (byte) 1);

    return new PetRecord(
        name, evs, ivs, talents, generation, breedCount, mountUnlocked, flightUnlocked);
  }

  private static double getDouble(PersistentDataContainer container, NamespacedKey key) {
    Double value = container.get(key, PersistentDataType.DOUBLE);
    return value != null ? value : 0.0;
  }

  private static byte asByte(boolean value) {
    return (byte) (value ? 1 : 0);
  }
}

File: src/main/kotlin/com/github/cybellereaper/wizpets/core/persistence/PetStorage.kt
package com.github.cybellereaper.wizpets.core.persistence

import com.github.cybellereaper.wizpets.api.PetRecord
import com.github.cybellereaper.wizpets.api.StatSet
import org.bukkit.NamespacedKey
import org.bukkit.entity.Player
import org.bukkit.persistence.PersistentDataContainer
import org.bukkit.persistence.PersistentDataType
import org.bukkit.plugin.java.JavaPlugin

class PetStorage(plugin: JavaPlugin) {
    private val rootKey = NamespacedKey(plugin, "pet")

    private val nameKey = NamespacedKey(plugin, "name")
    private val generationKey = NamespacedKey(plugin, "generation")
    private val breedKey = NamespacedKey(plugin, "breed_count")
    private val talentsKey = NamespacedKey(plugin, "talents")
    private val mountKey = NamespacedKey(plugin, "mount_unlocked")
    private val flightKey = NamespacedKey(plugin, "flight_unlocked")

    private val evKeys = mapOf(
        "health" to NamespacedKey(plugin, "ev_health"),
        "attack" to NamespacedKey(plugin, "ev_attack"),
        "defense" to NamespacedKey(plugin, "ev_defense"),
        "magic" to NamespacedKey(plugin, "ev_magic")
    )
    private val ivKeys = mapOf(
        "health" to NamespacedKey(plugin, "iv_health"),
        "attack" to NamespacedKey(plugin, "iv_attack"),
        "defense" to NamespacedKey(plugin, "iv_defense"),
        "magic" to NamespacedKey(plugin, "iv_magic")
    )

    fun load(player: Player): PetRecord? {
        val parent = player.persistentDataContainer
        val container = parent.get(rootKey, PersistentDataType.TAG_CONTAINER) ?: return null
        val name = container.get(nameKey, PersistentDataType.STRING) ?: return null
        val generation = container.get(generationKey, PersistentDataType.INTEGER) ?: 1
        val breedCount = container.get(breedKey, PersistentDataType.INTEGER) ?: 0
        val talents = container.get(talentsKey, PersistentDataType.STRING)?.split(';')?.filter { it.isNotBlank() } ?: emptyList()
        val mountUnlocked = container.get(mountKey, PersistentDataType.BYTE)?.toInt() == 1
        val flightUnlocked = container.get(flightKey, PersistentDataType.BYTE)?.toInt() == 1

        val evs = StatSet(
            health = container.get(evKeys.getValue("health"), PersistentDataType.DOUBLE) ?: 0.0,
            attack = container.get(evKeys.getValue("attack"), PersistentDataType.DOUBLE) ?: 0.0,
            defense = container.get(evKeys.getValue("defense"), PersistentDataType.DOUBLE) ?: 0.0,
            magic = container.get(evKeys.getValue("magic"), PersistentDataType.DOUBLE) ?: 0.0
        )
        val ivs = StatSet(
            health = container.get(ivKeys.getValue("health"), PersistentDataType.DOUBLE) ?: 0.0,
            attack = container.get(ivKeys.getValue("attack"), PersistentDataType.DOUBLE) ?: 0.0,
            defense = container.get(ivKeys.getValue("defense"), PersistentDataType.DOUBLE) ?: 0.0,
            magic = container.get(ivKeys.getValue("magic"), PersistentDataType.DOUBLE) ?: 0.0
        )

        return PetRecord(
            displayName = name,
            evs = evs,
            ivs = ivs,
            talentIds = talents,
            generation = generation,
            breedCount = breedCount,
            mountUnlocked = mountUnlocked,
            flightUnlocked = flightUnlocked
        )
    }

    fun save(player: Player, record: PetRecord) {
        val parent = player.persistentDataContainer
        val ctx = parent.adapterContext
        val container = ctx.newPersistentDataContainer()

        container.set(nameKey, PersistentDataType.STRING, record.displayName)
        container.set(generationKey, PersistentDataType.INTEGER, record.generation)
        container.set(breedKey, PersistentDataType.INTEGER, record.breedCount)
        container.set(talentsKey, PersistentDataType.STRING, record.talentIds.joinToString(";"))
        container.set(mountKey, PersistentDataType.BYTE, record.mountUnlocked.asByte())
        container.set(flightKey, PersistentDataType.BYTE, record.flightUnlocked.asByte())

        container.set(evKeys.getValue("health"), PersistentDataType.DOUBLE, record.evs.health)
        container.set(evKeys.getValue("attack"), PersistentDataType.DOUBLE, record.evs.attack)
        container.set(evKeys.getValue("defense"), PersistentDataType.DOUBLE, record.evs.defense)
        container.set(evKeys.getValue("magic"), PersistentDataType.DOUBLE, record.evs.magic)

        container.set(ivKeys.getValue("health"), PersistentDataType.DOUBLE, record.ivs.health)
        container.set(ivKeys.getValue("attack"), PersistentDataType.DOUBLE, record.ivs.attack)
        container.set(ivKeys.getValue("defense"), PersistentDataType.DOUBLE, record.ivs.defense)
        container.set(ivKeys.getValue("magic"), PersistentDataType.DOUBLE, record.ivs.magic)

        parent.set(rootKey, PersistentDataType.TAG_CONTAINER, container)
    }

    fun clear(player: Player) {
        player.persistentDataContainer.remove(rootKey)
    }

    private fun Boolean?.asByte(): Byte = if (this == true) 1 else 0
}
